independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java

```diff
@@ -505,10 +513,40 @@ public boolean test(MethodInfo method) {
                 // Do not skip default methods - public non-abstract instance methods declared in an interface
                 return false;
             }
+
+            List<Type> parameters = method.parameters();
+            if (!parameters.isEmpty() && (beanArchiveIndex != null)) {
+                String originalClassPackage = determinePackage(originalClazz.name());
+                for (Type type : parameters) {
+                    ClassInfo parameterClassInfo = beanArchiveIndex.getClassByName(type.name());
+                    if (parameterClassInfo == null) {
+                        continue; // hope for the best
+                    }
+                    if (Modifier.isPrivate(parameterClassInfo.flags())) {
+                        return true; // parameters whose class is private can not be loaded, as we would end up with IllegalAccessError when trying to access the use the load the class
+                    }
+                    if (!Modifier.isPublic(parameterClassInfo.flags())) {
+                        if (determinePackage(parameterClassInfo.name()).equals(originalClassPackage)) {
+                            return false;
+                        }
+                        // parameters whose class is package-private and the package is not the same as the package of the method for which we are checking can not be loaded,
+                        // as we would end up with IllegalAccessError when trying to access the use the load the class
+                        return true;
+                    }
+                }
+            }
+
             // Note that we intentionally do not skip final methods here - these are handled later
             return false;
         }
 
+        private String determinePackage(DotName dotName) {
+            if (dotName.isInner()) {
+                dotName = dotName.prefix();
+            }
+            return dotName.prefix() == null ? "" : dotName.prefix().toString();
```

Comment:
I think that this will not work for names created via `DotName.createSimple()`: _"Simple DotName variants never have a prefix"_.

Reply:
What would the proper use be in that case?

Reply:
`io.quarkus.arc.processor.DotNames.packageName(DotName)` should work although the impl is pretty "simplistic" :-)

Reply:
I was not aware of it, I'll put it on the TODO list :)